Core/Core/NetworkAvailability/NWAvailabilityService.swift

```diff
@@ -0,0 +1,86 @@
+//
+// This file is part of Canvas.
+// Copyright (C) 2023-present  Instructure, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+//
+
+import Combine
+import Foundation
+import Network
+
+public protocol NWAvailabilityService {
+    func startMonitoring()
+    func stopMonitoring()
+
+    func startObservingStatus() -> CurrentValueSubject<NWAvailabilityStatus, Never>
+    var status: NWAvailabilityStatus { get }
+}
+
+public final class NWAvailabilityServiceLive: NWAvailabilityService {
+    // MARK: - Depepdencies
+
+    private let monitor: NWPathMonitorWrapper
+
+    // MARK: - Properties
+
+    public private(set) var status: NWAvailabilityStatus = .disconnected {
+        didSet {
+            statusSubject.send(status)
+        }
+    }
+
+    private let statusSubject = CurrentValueSubject<NWAvailabilityStatus, Never>(.disconnected)
+    private var isMonitoring = false
+    private let queue = DispatchQueue(label: "com.instructure.icanvas.network-availability")
+
+    public init(monitor: NWPathMonitorWrapper = NWPathMonitorWrapper(from: NWPathMonitor())) {
```

Comment:
It seems that the NWPathMonitor can only handle a single instance requesting updates so we can't create a single app level instance of it and share it amongst multiple services because they would overwrite each other's updateHandlers and interfere with the start/stop method calls.
I would either add a comment to this init explaining this hidden limitation or add an assert to where the `updateHandler` is set that checks if the monitor already has an `updateHandler` set up. If yes, that's a programmer error and we can abort.

Reply:
Good idea, added a comment and an assertion.